idseq_dag/util/log.py

```diff
@@ -34,6 +39,70 @@ def write(message, warning=False, flush=True):
         if flush:
             sys.stdout.flush()
 
+def log_event(event_name, values=None, start_time=None, warning=False, flush=True):
+    '''
+    Write a new log event.
+
+    Parameters:
+    event_name (str): name of the event
+    values(dict): Optional. Values associated to that event. Will be logged in json format.
+    start_time(datetime): Optional. If given, it will calc the elapsed time from this datetime to now and write it to the log line.
+
+    Returns:
+    datetime: Now. It can be used to pass to parameter start_time in a future log_event call.
+
+    Example:
+    start = log_event("downloaded_started", {"file":"abc"})
+    # ... download your file here ...
+    log_event("downloaded_completed", {"file":"abc"}, start_time=start)
+    '''
+    fmt_values = "{}" if values is None else json.dumps(values)
+    if start_time is None:
+        fmt_message = "Event '%s' %s" % (event_name, fmt_values)
+    else:
+        duration = (datetime.datetime.now()-start_time).total_seconds()
+        fmt_message = "Event '%s' %s (%.1f seconds)" % (event_name, fmt_values, duration)
+    write(fmt_message, warning, flush)
+    return datetime.datetime.now()
+
+@contextmanager
+def log_context(context_name, values=None, log_caller_info=True):
+    '''
+    Log context manager to track started and completed events for a block of code
+
+    Parameters:
+    context_name (str): Name for this context.
+    values(dict): Optional. Values associated to that event. Will be logged in json format.
+    log_caller_info(bool): Log the caller file_name and method name.
+
+    Example:
+    # some_file.py
+    from log import log_context
+
+    def some_method
+        # ... some code here ...
+        with log_context("sub_step", {"abc": "123"}):
+            # ... your code goes here ...
+
+    The code above will generate two log entries:
+    Event ctx_start {"n": "sub_step", "v": {abc": 123}, "caller": {"f": ["some_file.py", "some_method"]}
```

Comment:
Why `n`, `v` vs. readable names? Just to save space?

Reply:
No, it was more related to declutter the visualization.
I had originally called them `context_name` and `values`, but I felt it created a lot of noise when reading these lines through CloudFront, which is our main tool to read these logs currently. Long names certainly wouldn't be a problem if we were fully ingesting these lines in JSON format and displaying in tabular format.

I also enjoy using descriptive names and I'm don't have a strong position on keeping them using short names. I'm very open to change it.

But I wanted a second opinion on how cluttered it actually gets.
Let me show an example:

### using short (and unreadable) names
```
2019-04-30 20:18:25,205: [    Thread-4] Command 9: date > /mnt/idseq/results/10589/banana_input.txt.done
2019-04-30 20:18:25,210: [    Thread-4] Event 'ctx_end' {"n": "fetch_input_files_from_s3", "v": {"target": "banana_in"}, "caller": {"f": ["pipeline_flow.py", "fetch_target_from_s3"]}} (1.4 seconds)
2019-04-30 20:18:28,849: [   Thread-11] Event 'ctx_end' {"n": "substep_wait_for_input_files", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,850: [   Thread-11] Event 'ctx_start' {"n": "substep_run", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,850: [   Thread-11] BANANA
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_end' {"n": "substep_run", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_start' {"n": "substep_validate", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,852: [   Thread-11] Command 10: date > /mnt/idseq/results/10589/banana1.txt.done
2019-04-30 20:18:28,858: [   Thread-11] Command 11: date > /mnt/idseq/results/10589/banana2.txt.done
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"n": "substep_validate", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"n": "substep_save_progress", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"n": "substep_save_progress", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"n": "substep_save_counts", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"n": "substep_save_counts", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"n": "dag_step", "v": {"step": "banana_out"}, "caller": {"f": ["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,863: [   Thread-16] Command 12: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana1.txt s3://sample-bucket/david/test_001/out/3.5/banana1.txt
2019-04-30 20:18:29,499: [   Thread-16] Command 13: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana2.txt s3://sample-bucket/david/test_001/out/3.5/banana2.txt
2019-04-30 20:18:30,173: [  MainThread] all steps are done
```

**vs**

### using descriptive names
```
2019-04-30 20:18:25,205: [    Thread-4] Command 9: date > /mnt/idseq/results/10589/banana_input.txt.done
2019-04-30 20:18:25,210: [    Thread-4] Event 'ctx_end' {"context_name":"fetch_input_files_from_s3", "values":{"target": "banana_in"}, "caller": {"function":["pipeline_flow.py", "fetch_target_from_s3"]}} (1.4 seconds)
2019-04-30 20:18:28,849: [   Thread-11] Event 'ctx_end' {"context_name":"substep_wait_for_input_files", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,850: [   Thread-11] Event 'ctx_start' {"context_name":"substep_run", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,850: [   Thread-11] BANANA
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_end' {"context_name":"substep_run", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,851: [   Thread-11] Event 'ctx_start' {"context_name":"substep_validate", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,852: [   Thread-11] Command 10: date > /mnt/idseq/results/10589/banana1.txt.done
2019-04-30 20:18:28,858: [   Thread-11] Command 11: date > /mnt/idseq/results/10589/banana2.txt.done
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"context_name":"substep_validate", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"context_name":"substep_save_progress", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_end' {"context_name":"substep_save_progress", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,862: [   Thread-11] Event 'ctx_start' {"context_name":"substep_save_counts", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}}
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"context_name":"substep_save_counts", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (0.0 seconds)
2019-04-30 20:18:28,863: [   Thread-11] Event 'ctx_end' {"context_name":"dag_step", "values":{"step": "banana_out"}, "caller": {"function":["pipeline_step.py", "thread_run"]}} (5.0 seconds)
2019-04-30 20:18:28,863: [   Thread-16] Command 12: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana1.txt s3://sample-bucket/david/test_001/out/3.5/banana1.txt
2019-04-30 20:18:29,499: [   Thread-16] Command 13: aws s3 cp --only-show-errors /mnt/idseq/results/10589/banana2.txt s3://sample-bucket/david/test_001/out/3.5/banana2.txt
2019-04-30 20:18:30,173: [  MainThread] all steps are done
```

Reply:
I like the second one, and think it will be better going forward and as we start analyzing these in data dog.